Improved BirdWeather Submissions

jaredb7 opened the pull request:
### **BirdWeather Submissions run in a thread**
Run BirdWeather submissions in a separate thread to run in the background it doesn't hold up or slow down processing audio and detections.

Previously if BirdWeather was slow or there were other network factors slowing down the submission, the analysis server would pause for the duration of the upload.... sometimes to a point of a backlog (I have experienced backlogs up to 5 minutes due to slow network conditions).

This enhancement allows the BirdWeather detections to added to a queue and processed in the background via a separate thread, this allow the analysis server to continue processing audio while sound scapes and detections upload at their own pace.

The single thread processing the queue also aids in accidentally sending too many BirdWeather submission at once, so we're only ever sending 1 submission at any one time.

Also greatly improved exception handling around requests, plus useful output in case things go wrong.
![image](https://user-images.githubusercontent.com/799998/235029091-aae227b0-6771-499a-8693-bf4188840021.png)




### **Request Timeouts, Retries & Back-off**

- [gerritzen](https://github.com/gerritzen/BirdNET-Pi/commits?author=gerritzen) Added timeout values to the requests as they will run indefinitely if no timeout is specified.

- A slightly longer timeout is used for the sound scape upload as this can take a while depending on network conditions. 60 seconds should be enough for a slow 3G cellular connection. And we want to give the sound scape the best chance to upload.

- Further, the ability to retry requests was implemented but only for connection timeouts (request didn't complete within our specified timeout period), connection errors (network problem (e.g. DNS failure, refused connection, etc)) and specific response codes that are known and can be retried on.
If a request raises a error which isn't a Connection Timeout or Connection Error, it will fall into the HTTP Error exception... we then check the response code to see what the error is and if it is any of the following it will be retried, anything else we consider a failure
```
404 Not Found
429 Too Many Requests
500 Internal Server Error
502 Bad Gateway
503 Service Unavailable
504 Gateway Timeout
```

The Sound scape upload and detection upload are retried separately, the sound scape may succeed but the detection might not, so the detection will get retried until succeeding or exhausting retries.
It's more likely the sound scape upload will be the source a issue as it's the most intensive of the two
Submissions will only try 6 times to upload, time-outs together with the back-off will see this happen over multiple minutes.

- A exponential back-off delay is calculated before retrying again to provide adequate delay between between requests and ease API server pressure.

```
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 1.7s, Retry Attempt:1 - TOO MANY REQUESTS
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 3.4s, Retry Attempt:2 - TOO MANY REQUESTS
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 6.8s, Retry Attempt:3 - TOO MANY REQUESTS
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 13.6s, Retry Attempt:4 - TOO MANY REQUESTS
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 27.2s, Retry Attempt:5 - TOO MANY REQUESTS
BirdWeather Submission:: Soundscape POST - HTTP Request Exception for <bird_weather>, Retrying after 54.4s, Retry Attempt:6 - TOO MANY REQUESTS
```

A working example where the retry system helped submit the detection 
```
May  1 07:00:44 BirdNET python3[2232720]: BirdWeather Submission:: Soundscape Successfully Uploaded - status:201 soundscape_id:4XX155XX
May  1 07:00:44 BirdNET python3[2232720]: BirdWeather Submission:: Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:00:45 BirdNET python3[2232720]: BirdWeather Submission Error:: Detection POST - HTTP Request Exception for https://app.birdweather.com/api/v1/stations/AxBxC/soundscapes?timestamp=2023-05-01T07:00:03+10:00, Retrying after 1.7s, Retry Attempt:1
May  1 07:00:47 BirdNET python3[2232720]: BirdWeather Submission:: Retry (1) - Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:00:48 BirdNET python3[2232720]: BirdWeather Submission Error:: Detection POST - HTTP Request Exception for https://app.birdweather.com/api/v1/stations/AxBxC/soundscapes?timestamp=2023-05-01T07:00:03+10:00, Retrying after 3.4s, Retry Attempt:2
May  1 07:00:51 BirdNET python3[2232720]: BirdWeather Submission:: Retry (2) - Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:00:52 BirdNET python3[2232720]: BirdWeather Submission Error:: Detection POST - HTTP Request Exception for https://app.birdweather.com/api/v1/stations/AxBxC/soundscapes?timestamp=2023-05-01T07:00:03+10:00, Retrying after 6.8s, Retry Attempt:3
May  1 07:00:59 BirdNET python3[2232720]: BirdWeather Submission:: Retry (3) - Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:00:59 BirdNET python3[2232720]: BirdWeather Submission Error:: Detection POST - HTTP Request Exception for https://app.birdweather.com/api/v1/stations/AxBxC/soundscapes?timestamp=2023-05-01T07:00:03+10:00, Retrying after 13.6s, Retry Attempt:4
May  1 07:01:13 BirdNET python3[2232720]: BirdWeather Submission:: Retry (4) - Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:01:14 BirdNET python3[2232720]: BirdWeather Submission Error:: Detection POST - HTTP Request Exception for https://app.birdweather.com/api/v1/stations/AxBxC/soundscapes?timestamp=2023-05-01T07:00:03+10:00, Retrying after 27.2s, Retry Attempt:5
May  1 07:01:41 BirdNET python3[2232720]: BirdWeather Submission:: Retry (5) - Detection Data - json: {"timestamp": "2023-05-01T07:00:06+10:00", "lat": "-23.XXXX", "lon": "150.XXXX", "soundscapeId": "4XX155XX", "soundscapeStartTime": "3.0", "soundscapeEndTime": "6.0", "commonName": "Magpie-lark", "scientificName": "Grallina cyanoleuca", "algorithm": "alpha", "confidence": "0.7420777"}
May  1 07:01:43 BirdNET python3[2232720]: BirdWeather Submission:: Retry (5) - Detection Successfully Uploaded - status:201 - 07:00:06/Magpie-lark/0.7420777

```

Any errors can be monitor for with grep as the Birdnet log can get a little busy
`sudo grep "BirdWeather" /var/log/syslog`

Closes #863

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                        print(
+                            f"BirdWeather Submission Error:: {retry_text}Detection POST - Unable to read response, something went wrong! - {idx_exec}",
+                            flush=True)
+                    except requests.exceptions.ConnectTimeout as conn_timeout_exec:
```
here until 968: see above (covered by for, no need for individual treatment of these errors)

jaredb7 replied:
some of these are gone now with unneeded response code removed (including IndexError). 918 is the similar to the original code with the extra debug line above it.

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                                flush=True)
+                    except requests.exceptions.RequestException as request_exc:
+                        # Check if the status code is one that we can try on
+                        req_status_code = request_exc.response.status_code
```
Is this not detection_response_status_code? Isn't this already covered above?

jaredb7 replied:
This is just checking the response code once we're in exception, it's separate from the check in the try body for 2XX status codes.. which isn't required

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                            print(
+                                f"BirdWeather Submission Error:: Detection POST - HTTP Request Exception for {soundscape_url}, Not retrying - Retries Attempted:{submission_retries} - {request_exc}",
+                                flush=True)
+                    except (
```
related to the comment above. If you agree that we don't need to decode the response json, then we don't need to treat this here.

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                            if request_backoff_time > 0:
+                                time.sleep(request_backoff_time)
+                                continue
+                else:
```
Can you turn this around maybe? This belongs to an if almost 200 lines ago. 
```
if detection_url is None or detection_post_json is None or soundscape_id is None:
    # handle this
else:
    # The 200 lines above
```

jaredb7 replied:
I removed both the if statements around section so the cascading else towards has been removed too

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                        f"BirdWeather Submission Error:: {retry_text}Detection POST did not succeed, detection_url, detection_json or soundscape_id missing or not supplied",
+                        flush=True)
+
+        else:
```
same as 5 lines above.

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+            print("BirdWeather Submission Error:: Soundscape POST did not succeed, detection was not submitted",
+                  flush=True)
+
+    else:
```
same.

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+        print(f"birdweather_soundscapes - mode:{mode} - for {soundscape_filename_to_find}")
+
+    # Default mode to search
+    if mode is None:
```
The comment doesn't explain this, but an explanation of "mode" would be nice. Also,
`def birdweather_soundscapes(mode='search', soundscape_filename_to_find, ...` would do the same as this if clause

jaredb7 replied:
It was going to be a function you could call without the mode to do a search but worked out differently.
Removed 1061

there's a small snippet in each in the if and elif that mentions briefly what the following code

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+        time.sleep(1)
+
+
+def birdweather_soundscapes(mode, soundscape_filename_to_find, bw_soundscape_id_for_filename=None):
```
Can you give this function a more meaningful name?

jaredb7 replied:
changed to birdweather_soundscape_id_cache

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+            # Get the BirdWeather submission data, this is a dictionary containing the necessary data
+            bw_submission_data = bw_submission_queue.get()
+
+            # Verify we have data
```
Doesn't line 1034 ensure this already?

jaredb7 replied:
1304 just checks the actual if the queue has items in it, else calling .get() on a empty queue will just block. 1039 isn't really required though. removed.

gerritzen reviewed (changes requested):
I think your code could do the same in like half the number of lines.

Maybe it's too late, maybe it's the hundreds of lines of nested indentation, but I'm not quite sure how two of the main problems are treated:

1. How do you avoid race conditions? (Try to upload detection number 2 of one soundscape before number 1 is uploaded)
2. How many threads do you spawn? You mentioned that you were able to launch up to 4, but I probably missed it. Or is it just one thread that works on the queue of submissions/soundscape uploads? If so, Doesn't this potentially create an enormous backlog if you wait something like 2 minutes per upload?

jaredb7 commented:
> * How do you avoid race conditions? (Try to upload detection number 2 of one soundscape before number 1 is uploaded)
The soundscape upload and detection are a single item in the queue and processed as single block, once that completes processes the next queue item.

> * How many threads do you spawn? You mentioned that you were able to launch up to 4, but I probably missed it. Or is it just one thread that works on the queue of submissions/soundscape uploads? If so, Doesn't this potentially create an enormous backlog if you wait something like 2 minutes per upload?

Just the 1 thread spawned now, a backlog could eventuate if the soundscape upload is constantly timing out or if the user is on a connection with weak upload bandwidth (the time-out is set to try accommodate this). Outside of that the max time the retry loops sleep is 54 seconds at the last (6th) retry, this can changed around but it's more or less same method the URLib Retry class uses


I have considered a backlog occurring but not sure how you'd prevent it if you want also give the uploads the very best chance of getting out.

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    soundscape_id = soundscape_response_json['soundscape']['id']
+
+                    # Print a success message
+                    if soundscape_response_status_code == 200 or soundscape_response_status_code == 201 or soundscape_response_status_code == 202:
```
correct, just leftover code, this and the following else have been removed

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+
+                        # Cache the soundscape id for corresponding to the filename the wave uploaded
+                        birdweather_soundscapes('add', soundscape_filename, soundscape_id)
+                    else:
```
removed with the above

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    # We found the soundscape filename and the Birdweather Soundscape ID for it, soundscape considered uploaded & use the ID for the detection
+                    soundscape_uploaded = True
+                    soundscape_id = find_existing_soundscape['soundscape_id']
+                    # if debug_birdweather_submissions:
```
Yeah this doesn't need to be output, the if statement above it should have been uncommented so it's output during debug how it was intended

The journal issue does need to be looked into, after clearing mine initially it hasn't grown as aggressively as it initially was. Could modify it's config and cap it, but it'd be more more beneficial sorting out where it's coming from.

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                        f"BirdWeather Submission:: {retry_text}Found Existing Soundscape in cache for {soundscape_filename} - using soundscape_id:{soundscape_id}",
+                        flush=True)
+
+            except IndexError as idx_exec:
```
Access a invalid array index will throw this, it was around from a earlier iteration, definitely not required

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    flush=True)
+            except requests.exceptions.ConnectTimeout as conn_timeout_exec:
+                # Check if were still able to retry
+                if submission_retries < bw_request_max_retries:
```
`can_retry` was just used a flag to determine if should sleep for the `request_backoff_time ` (which occurs as a  last step)

continue together moving the calculation and sleep around should work

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+            except requests.exceptions.RequestException as request_exc:
+                # Check if the status code is one that we can try on
+                req_status_code = request_exc.response.status_code
+                if req_status_code in bw_request_status_retry_for:
```
First time checking the response code once we're in exception, it's separate from the unneeded check in the try body for 2XX response and is just so we know if we're in a state of being able to retry.

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    if submission_retries < bw_request_max_retries:
+                        can_retry = True
+                        submission_retries += 1
+                        request_backoff_time = bw_request_backoff_factor * (2 ** (submission_retries - 1))
```
I've removed this and others and just have the calculation & sleep done at the begin retry loop

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    print(
+                        f"BirdWeather Submission Error:: Soundscape POST - HTTP Request Exception for {soundscape_url}, Not retrying - Retries Attempted:{submission_retries} - {request_exc}",
+                        flush=True)
+            except (requests.exceptions.JSONDecodeError, requests.exceptions.InvalidJSONError) as json_error_exec:
```
Some is redundant I agree but some additional context around errors helps in quickly identifying where it fell. definitely can be reduced some

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                    break
+                else:
+                    # We're retrying, retry after the calculated backoff time
+                    if request_backoff_time > 0:
```
no not required actually but it was being reset within the respective retry loops

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+                        continue
+
+        # Soundscape has been uploaded, now we can submit the detection data to birdnet
+        if soundscape_uploaded:
```
We can reach here if the retry loop for the soundscape upload completes without uploading (kept timing out for example) thus no soundcape id. Then don't attempt to upload the detection if that's the case.

jaredb7 commented on scripts/server.py:
```diff
@@ -550,66 +567,544 @@ def handle_client(conn, addr):
+
+                        # Check the response
+                        if detection_response_status_code == 200 or detection_response_status_code == 201 or detection_response_status_code == 202:
+                            # Extract the bird detection info to display in the output
```
It is just parroting what was just sent but also sanity checks/confirms birdweather processed what we sent. There's a very slight difference in what's returned with valid vs invalid data (unlikely).

This is just from observation but if it processed the detection correctly then json data with  the key 'detection'  representing the detection sent, else returns all today's detections under 'detections' none of which are the detection I posted.

It isn't needed but I also don't think it hurts having some visual cue of the uploaded detection.

jaredb7 commented:
@gerritzen I've rearranged and trimmed of any unneeded/left in by mistake code and run a few days just to be sure, looking better. do you have any further  suggestions?

jaredb7 committed:
Threaded BirdWeather submissions with retry support

Run BirdWeather submissions in a separate thread to run in the background it doesn't hold up or slow down processing audio and detection's.

Instead of creating a new thread to process each Birdweather submission as they occur, add the data used for the Birdweather submission to a queue.

Then use a single thread running birdweather_submission_processor in a endless loop to process Birdweather submissions as they are added to that queue.

This avoids accidentally sending too many Birdweather submission at once, so we're only ever sending 1 submission at any one time.

Add timeout values to the requests as they will run indefinitely if no timeout is specified

Further, the ability to retry requests was implemented but only for connection timeouts and specific responses codes that can be retried on.

A exponential backoff delay is calculated before retrying again to provide adequate delay between requests and ease server pressure

gerritzen commented on scripts/server.py:
```diff
@@ -550,66 +573,329 @@ def handle_client(conn, addr):
+                continue
+            else:
+                print(f"BirdWeather Submission Error:: Soundscape POST - HTTP Request Exception! Cannot retry - {request_exc}", flush=True)
+                # break the loop on non-retryable status
```
Not sure if that's a word, but if it is, it's spelled "retriable" ;)

jaredb7 replied:
Yeah it's definitely not a common place compound word but gets used here and there, I've mostly seen it around Microsoft Azure AD SSO integration and SQL Server errors.

Either "retryable" or "retriable" are correct spelling for basically the same word, though either way might be suitable depending on the context.

gerritzen reviewed (commented):
Looks a lot better! I haven't tested it, but apart from the "2p2" I saw no obvious error. If you say that it runs stably, then I'd give my thumbs up.

jaredb7 committed:
Updated to include changes in commit #479bd27 & BW improvement added
